src/demangle_codewarrior.py

```diff
@@ -0,0 +1,460 @@
+            elif names[-1] == "__dt":
+                assert len(names) >= 2
+                names[-1] = f"~{names[-2]}"
+            return "::".join(names)
+
+        if self.kind == CxxTerm.Kind.FUNCTION:
+            assert self.function_params is not None
+            prefix = ""
+            if self.function_return is not None:
+                prefix = f"{self.function_return} (*) "
+            args = ", ".join(str(p) for p in self.function_params)
+            return f"{prefix}({args})"
+
+        if self.kind == CxxTerm.Kind.SYMBOL_REFERENCE:
+            assert self.symbol_reference is not None
+            return f"&{self.symbol_reference}"
+
+        if self.kind == CxxTerm.Kind.POINTER:
+            return "*"
+        if self.kind == CxxTerm.Kind.REFERENCE:
+            return "&"
+        if self.kind == CxxTerm.Kind.LONG_LONG:
+            return "long long"
+        if self.kind == CxxTerm.Kind.LONG_DOUBLE:
+            return "long double"
+        if self.kind == CxxTerm.Kind.WIDE_CHAR:
+            return "wchar_t"
+        if self.kind == CxxTerm.Kind.ELLIPSIS:
+            return "..."
+        return self.kind.name.lower()
+
+
+@dataclass
+class CxxType:
+    terms: List[CxxTerm] = field(default_factory=list)
+
+    @staticmethod
+    def parse(src: TextIOBase) -> "CxxType":
+        terms = []
+        while True:
+            if peek(src) in (",", ">"):
+                break
+            term = CxxTerm.parse(src)
+            terms.append(term)
+            if term.kind not in CxxTerm.NONTERMINATING_KINDS:
+                break
+        return CxxType(terms=terms)
+
+    def __str__(self) -> str:
+        return " ".join(str(a) for a in self.terms[::-1]).strip()
+
+
+@dataclass
+class CxxSymbol:
+    STATIC_FUNCTIONS = {"__sinit", "__sterm"}
+
+    name: CxxTerm
+    type: CxxType
+
+    @staticmethod
+    def parse(src: TextIOBase) -> "CxxSymbol":
+        with peeking(src):
+            base_name = src.read()
+
+        strip_underscores = 0
+        template_depth = 0
+        with peeking(src):
+            chars = ""
+            while True:
+                c = src.read(1)
+                if not c:
+                    break
+
+                if c in (",", ">") and template_depth == 0:
+                    break
+                elif c == "<":
+                    template_depth += 1
+                elif c == ">":
+                    template_depth -= 1
+
+                chars += c
+                if chars in CxxSymbol.STATIC_FUNCTIONS:
+                    base_name = chars
+                    strip_underscores = 1
+                    break
+                elif not (c == "_" and peek(src) == "_"):
+                    continue
+                strip_underscores = 2
+                lookahead = peek(src, 2)
+                if len(lookahead) == 2:
+                    if lookahead[1] not in "QFC0123456789":
+                        continue
+                base_name = chars[:-1]
```

Comment:
This still makes little sense to me, even after the new comments: "Peek at the char after the "__": if it could be the start of the mangled type info, do not advance the `base_name`." but why? if the mangled type info starts there, isn't that exactly where you'd want to allow a base_name break point? I'm also confused as to what `QFC0123456789` represents exactly. Examples would help here I suspect.

(nit: the two ifs can be folded into one with an `and`)

Also, all this special treatment of underscores makes we wonder if that's related to C++'s rules on symbol names:

> [C++03] 17.4.3.1.2 Global names [lib.global.names]
> 
> Certain sets of names and function signatures are always reserved to the implementation:
> 
> * Each name that contains a double underscore (__) or begins with an underscore followed by an uppercase letter (2.11) is reserved to the implementation for any use.
> * Each name that begins with an underscore is reserved to the implementation for use as a name in the global namespace.

Haven't associated this with name mangling before, but it makes a certain amount of sense. Not something to rely on if we can avoid it though, with those rules mostly being treated as loose guidelines by programmers.

Reply:
Heh, I didn't realize double underscore is disallowed!

Unfortunately, I think there are a few examples in SMS of symbols that violate this rule, like `__start__Q27JAInter9StreamLibFv` (`JAInter::StreamLib::__start (void)`) 

I've refactored it to remove the double-negative (`if not ...: continue`), re-wrote the comment, and added some tests that are more explicitly for this loop. I also explicitly called out that this is a heuristic, and highlighted some shortcomings. 

Let me know if there's something else I could add to make this more clear?

Reply:
Fantastic! Hearing that it's a heuristic explains a lot.